**Importing the Dependices**

In [ ]:
#libraries 
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.metrics import accuracy_score 

In [ ]:
# Dataset in csv format (coma separated values)
# loading the csv data to pandas dataframe 
# Kaggle Data set link
# https://www.kaggle.com/datasets/johnsmith88/heart-disease-dataset

heart_data = pd.read_csv('/content/heart.csv')

In [ ]:
# Printing the first five rows of the dataframe  
heart_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [ ]:
# Printing the Last five rows of the dataframe  
heart_data.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [ ]:
# No. of rows and column in dataset 
heart_data.shape

(303, 14)

**Insights :** Small Data set , Logistic Regression can perform well in this case 

Getting Information for **checking of the missing and the null** values in the dataset 

In [ ]:
# getting some info about the data
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
# checking for missing values 
heart_data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

If missing values Present,handle the missing values using imputation

Statistical Measures about the data

In [ ]:
heart_data.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

**Insights :** 25% means 25 /5 of the value are less than 47.5(in case of age ) 

**Checking the Distribution**

**1 -> Defective Heart**

**0 -> Healthy Heart**

In [ ]:
# Checking the distribution of target variable 
heart_data['target']
heart_data['target'].value_counts()

1    165
0    138
Name: target, dtype: int64

**Splitting the features and target**

In [ ]:
# Dropping a column that's why axis = 1 
X = heart_data.drop(columns = 'target',axis = 1)
Y = heart_data['target']

In [ ]:
print(X)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0    

In [ ]:
print(Y)

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64


**Train Test Split (Splitting the Training and Testing Data)**


In [ ]:
# Stratify is used to check that proper division of data is done and division in even manner 
# if not used then it might lead that training data contains all 0's and test data contains all 1's 
# test size = 0.2 -> 20 % of data used for testing 
# random state represents data is distributed randomly
 
X_train,X_test,Y_train,Y_test = train_test_split(X, Y, test_size=0.2, stratify = Y,random_state = 2)

In [ ]:
# Printing the number of data spilts 
print(X.shape, X_train.shape, X_test.shape)

(303, 13) (242, 13) (61, 13)


**Model Training**

**Logistic Regression**

* Mainly used for **Binary Classifcation Problems**
* Works Perfectly Well for Small Dataset
* **No issue of Overfitting** on Small Dataset

**Overfitting :** 
Training data predicts very well
Testing data predicts very poorly

In [ ]:
# model = LogisticRegression()
model = svm.SVC(kernel='linear')

In [ ]:
# training the Logistic Regression Model with training data
model.fit(X_train, Y_train) 

SVC(kernel='linear')

**Model Evaluation**

Accuracy Score

Accuracy Score on Training Data

In [ ]:
# Accuracy on Training Data 
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train) 

In [ ]:
print('Accuracy on Training data: ', training_data_accuracy)

Accuracy on Training data:  0.8553719008264463


Accuracy Score on Training Data

In [ ]:
# Accuracy on Test Data 
X_test_prediction = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test) 

In [ ]:
print('Accuracy on Test data: ', testing_data_accuracy)

Accuracy on Test data:  0.819672131147541


**Building a Predictive System**

In [ ]:
# to get the input values 

# tuple data type 
# input_data = (60,0,0,150,258,0,0,157,0,2.6,1,2,3)
input_data = (63,1,3,145,233,1,0,150,0,2.3,0,0,1)

# changing the data to a numpy array as computation is easier in numpy array 
input_data_as_numpy_array = np.asarray(input_data)

# Reshaping as we want to predict only for one value at a time 
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = model.predict(input_data_reshaped)
print(prediction)

# prediction is in the form of list thats why using prediction[0]
if (prediction[0] == 0):
  print('The Person does not have a Heart Disease')
else:
  print('The Person has Heart Disease')

[1]
The Person has Heart Disease


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  "X does not have valid feature names, but"


**Saving the Trained Model**

In [ ]:
import pickle

In [ ]:
filename = 'heart_disease1.sav'
pickle.dump(model,open(filename,'wb'))

In [ ]:
# loading the saved model
loaded_model = pickle.load(open('/content/heart_disease1.sav','rb'))

**Making Predictive System using the loaded model**

In [ ]:
# to get the input values 

# tuple data type 
#input_data = (60,0,0,150,258,0,0,157,0,2.6,1,2,3)
input_data = (63,1,3,145,233,1,0,150,0,2.3,0,0,1)

# changing the data to a numpy array as computation is easier in numpy array 
input_data_as_numpy_array = np.asarray(input_data)

# Reshaping as we want to predict only for one value at a time 
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = loaded_model.predict(input_data_reshaped)
print(prediction)

# prediction is in the form of list thats why using prediction[0]
if (prediction[0] == 0):
  print('The Person does not have a Heart Disease')
else:
  print('The Person has Heart Disease')

[1]
The Person has Heart Disease


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  "X does not have valid feature names, but"
